### ISSUE TITLE
need to be able to image a range of channels, and average channels together

### ISSUE THREAD

**Issue by o-smirnov** [ska-sa/mt-imager#6]
We need to be able to 
- image a subset of channels (e.g. 1:63 meaning channels 1 to 62, or e.g. 0:59:2 meaning every other channel i.e. 0, 2, 4, ... 58.
- average channels together. A common use case is to average all channels before gridding (i.e. all channels in the specified subset) and produce a single-frequency FITS file. Another common one is to average every N channels together. Need both options.

**Comment by ddmusc** [comment]
@o-smirnov, I am working to enable such enhancement but I need ideas to how the configuration format should be I am thinking of using a CASA style syntax such as spw=0:5~10 but notsure if the best way

**Comment by twillis449** [comment]
Nothing wrong with CASA syntax for this stuff. In fact if would be nice if you followed CASA syntax where possible for specifications. That way one wouldn't have to remember how to specify the same thing for different imaging programs.

**Comment by o-smirnov** [comment]
Yeah I agree with @twillis449. The CASA syntax seems fine, no point in breeding new syntaxes. @IanHeywood might have some good ideas too?

**Comment by ddmusc** [comment]
But how then with CASA you say you want to average out two channels together?

**Comment by o-smirnov** [comment]
I think you either specify the number of channels in the output image, or the averaging "step". You then subdivide whatever channels were selected by the input syntax into that number of output channels. Of course this leaves you with all sorts of boundary cases where the output frequency axis is no longer regular. In these cases I say print a warning and do something reasonable (the FITS standard has something called imageIndex to support this).

**Comment by twillis449** [comment]
I may be wrong, but I don't think CASA 'averages' channels together. The default method of imaging is to use MFS, which as far as I'm aware grids each channel at its correct wavelength/frequency in UV space. You specify the range of channels etc that end up in a particular image by specification with the 'spw' keyword. 

Since you are modifying things, examples of how things are specified in current CASA-land are shown if you start up casapy and do 'help clean'. After wading through numerous screens of output you eventually get to an 'Example :' section which gives detailed examples of how the CASA user can specify data selection via spw, field, timerange, uvrange, mfs  etc etc. etc.

**Comment by twillis449** [comment]
@o-smirnov  CASA lets you 'account' for spectral index variations by using Taylor term expansion to model the frequency dependence of sky emission. The default is nterms=1, which means don't model anything.

**Comment by o-smirnov** [comment]
Yeah sure, my point about an irregular frequency axis was representing that in the FITS file. The normal CRVAL/CRPIX/CDELT way assumes a regular grid. If the user specifies some awkward combination of spectral windows and channel subsets, you may end up needing to produce an irregularly sampled freq axis. Hence imageIndex and all that.

Also, even if CASA doesn't "average" channels together anymore (lwimager still has that mode, one can select either plain averaging or MFS, right?), but in any case that's an orthogonal issue. @ddmusc needs some interface to specify that some number of input channels, across some number of spectral windows, ends up mapped into N output channels. Whether that's done with MFS or some other algorithm internally is a separate issue (Cyril and Andre have various interesting alternative ideas) and should be controlled by separate parameters. The general channel selection interface of "I want to take this subset of the input channels, and I want to produce these output channels" doesn't depend on the algorithm details (I think....)

**Comment by ddmusc** [comment]
That's my aim... I was thinking the best is use the spw syntax... and use
another variable step or channelAdd.. where you define how much selected
channels you are to add in one image. If in your definition it comes out
that you have an irregular frequency axis then the mt-imager will fail
stating the problem.

I believe it is mostly what you are arguing with the exception of irregular
axis.. I don't see the advantage of trying to image also in the irregular
axis case since the user can just run the imager more than once to get the
required behaviour

Goodnight... falling asleep on my chair:)

**Comment by twillis449** [comment]
The spw keyword lets you specify irregular groups of channels which end up collected into one image via mfs.  The spw parameters can be very complicated - see the examples area in 'help clean' that I mentioned above. If you are observing in channel mode, you can specify a 'width' parameter which will allow you to condense down the number of images in your output spectral cube, BUT the stepping intervals still have to be regular - I don't think that FITS has any specifications for irregular steps in CDELT space. As Daniel says I think if you have some weird irregular steps then you have to re-run the imager.

Lucky me - for SKA simulations I only use one channel so I don't need to worry about the problems being discussed here. The mind boggles at the amount if computation that would be required for simulations with more than one channel!
